### PR TITLE
use a separate bind-time flag instead of FI_COMPLETION

### PR DESCRIPTION
**Pull request by goodell**
Per discussion in issue #954, add a new flag (`FI_SELECTIVE_COMPLETION`) for bind-time usage to avoid overloading the meaning of `FI_COMPLETION` in a confusing manner.

I have a few notes about this PR:
- The sockets provider defines `SOCK_NO_COMPLETION`, which happens to have the same value as the new `FI_SELECTIVE_COMPLETION` flag.  I can't tell yet if this aliasing causes any problems, but it passes tests on my machines so I think it's fine.
- I changed the names of the `send_cq_event_flag`/`recv_cq_event_flag` variables in the psm provider, but I can change it back if the old name or some other name is actually preferred.
- I included a small commit to drop `FI_COMPLETION` from the sockets provider's capabilities flags.  If that's an incorrect change then I think that the man pages and some of the other providers need to be updated instead.

@shefty @jithinjosepkl @shantonu @a-ilango @pmmccorm please check that I haven't mangled your respective providers' code in an unacceptable manner

### CODE REVIEW

**Comment by jithinjosepkl** [comment]
Can we pick another value for FI_SELECTIVE_COMPLETION? 
60-63 are provider specific as per fabric.h

**Comment by goodell** [comment]
> Can we pick another value for FI_SELECTIVE_COMPLETION? 60-63 are provider specific as per fabric.h

Sure.  I missed that comment.  I'll respin in a second.

**Comment by goodell** [comment]
Fixed. It uses `(1ULL << 55)` now instead.  Thanks for pointing out that comment.

File: include/rdma/fabric.h [hunk 1]
```diff
@@ -121,6 +121,7 @@ typedef struct fid *fid_t;
 #define FI_TRANSMIT_COMPLETE	(1ULL << 27)
 #define FI_DELIVERY_COMPLETE	(1ULL << 28)
 
+#define FI_SELECTIVE_COMPLETION	(1ULL << 55)
```
Comment (shefty) [inline include/rdma/fabric.h]: This is a bind specific call flag, but it is being grouped with the fi_getinfo specific caps.  Please move the definition closer to struct fi_ops, where bind is defined, and change the value to 59.  FI_SELECTIVE_COMPLETE is not used with fi_getinfo.

File: man/fi_endpoint.3.md [hunk 2]
```diff
@@ -256,42 +256,58 @@ together when binding an endpoint to a completion domain CQ.
...
+
+  {% highlight c %}
+  fi_tx_attr::op_flags = 0;
+  fi_ep_bind(ep, cq, FI_SEND);  // default - completion
+  fi_send(ep, ...);                   // completion
+  fi_sendv(ep, ...);                  // completion
+  fi_sendmsg(ep, ..., 0);             // completion!
+  fi_sendmsg(ep, ..., FI_COMPLETION); // completion
+  fi_sendmsg(ep, ..., FI_INJECT|FI_COMPLETION); // completion!
+  fi_inject(ep, ...);                 // no completion!
```
Comment (shefty) [inline man/fi_endpoint.3.md]: :+1:

**Comment by shefty** [comment]
The socket and psm provider changes look good to me, but tagging @jithinjosepkl and @j-xiong to confirm the changes.

File: include/rdma/fabric.h (hunk 1 above)
Comment (goodell) [inline include/rdma/fabric.h]: I'm clearly very confused by this grouping/numbering scheme.  59 is already in use by `FI_DIRECTED_RECV` and 60 is reserved for provider-specific values.  `FI_SEND` is passed into `fi_bind` in the same bind flags argument as `FI_SELECTIVE_COMPLETION` but it has a value of 13.  I can make the change that you are requesting, but I'm baffled at how I would arrive at that decision on my own...

Comment (goodell) [inline include/rdma/fabric.h]: Oh, I think I see it now.  The `59-go down` space is reused multiple times for different classes.  That was... not intuitive to me.

**Comment by jithinjosepkl** [comment]
Sockets provider changes look good to me.

**Comment by j-xiong** [comment]
The PSM provider changes look fine to me.

File: include/rdma/fabric.h (hunk 1 above)
Comment (shefty) [inline include/rdma/fabric.h]: See this comment near the top of fabric.h:

/*
 * Flags
 * The 64-bit flag field is used as follows:
 * 1-grow up    common (usable with multiple operations)
 * 59-go down   operation spceific (used for single call/class)
 * 60 - 63      provider specific
 */

complete with a misspelling of specific.  If you're still in the code, can you fix the spelling at part of the same commit?

Comment (goodell) [inline include/rdma/fabric.h]: > See this comment near the top of fabric.h:

Right, I see now what that comment was attempting to convey, but it didn't match up with how I was thinking about the flag space (my natural desire was to assign unique values for every single flag with no reuse as long as there were still bits available).

> If you're still in the code, can you fix the spelling at part of the same commit?

Yes, already done as a separate tiny commit that I'll push in a moment.  I was just waiting to see if any other comments came in before pushing.

**Comment by goodell** [comment]
OK, PR updated with 59 change and a small flags clarification commit.

**Comment by shefty** [comment]
thanks!

**Comment by a-ilango** [comment]
Verbs provider changes look fine. Thanks!

**Pull request merged**